src/vs/workbench/parts/files/browser/views/explorerViewer.ts

```diff
@@ -526,20 +526,53 @@ export class FileController extends DefaultController {
 
 // Explorer Sorter
 export class FileSorter implements ISorter {
+	private toDispose: IDisposable[];
+	private sortOrder: SortOrder;
+
+	constructor(
+		@IConfigurationService private configurationService: IConfigurationService
+	) {
+		this.toDispose = [];
+
+		this.onConfigurationUpdated(configurationService.getConfiguration<IFilesConfiguration>());
+
+		this.registerListeners();
+	}
+
+	private registerListeners(): void {
+		this.toDispose.push(this.configurationService.onDidUpdateConfiguration(e => this.onConfigurationUpdated(this.configurationService.getConfiguration<IFilesConfiguration>())));
+	}
+
+	private onConfigurationUpdated(configuration: IFilesConfiguration): void {
+		this.sortOrder = configuration && configuration.explorer && configuration.explorer.sortOrder || 'default';
+	}
 
 	public compare(tree: ITree, statA: FileStat, statB: FileStat): number {
-		if (statA.isDirectory && !statB.isDirectory) {
-			return -1;
-		}
+		switch (this.sortOrder) {
```

Comment:
To implement the approach you described and keep current behavior of placeholders, the code may look like this:
```ts
public compare(tree: ITree, statA: FileStat, statB: FileStat): number {
	// Do not sort roots
	if (statA.isRoot) {
		return -1;
	}
	if (statB.isRoot) {
		return 1;
	}

	switch (this.sortOrder) {
		case 'default':
			if (statA.isDirectory && !statB.isDirectory) {
				return -1;
			}
			if (statB.isDirectory && !statA.isDirectory) {
				return 1;
			}
			if (statA instanceof NewStatPlaceholder) {
				return -1;
			}
			if (statB instanceof NewStatPlaceholder) {
				return 1;
			}
			return comparers.compareFileNames(statA.name, statB.name);

		case 'type':
			if (statA.isDirectory && !statB.isDirectory) {
				return -1;
			}
			if (statB.isDirectory && !statA.isDirectory) {
				return 1;
			}
			if (statA instanceof NewStatPlaceholder) {
				return -1;
			}
			if (statB instanceof NewStatPlaceholder) {
				return 1;
			}
			return comparers.compareFileExtensions(statA.name, statB.name);

		case 'modified':
			if (statA.isDirectory && !statB.isDirectory) {
				return -1;
			}
			if (statB.isDirectory && !statA.isDirectory) {
				return 1;
			}
			if (statA instanceof NewStatPlaceholder) {
				return -1;
			}
			if (statB instanceof NewStatPlaceholder) {
				return 1;
			}
			if (statA.mtime !== statB.mtime) {
				return statA.mtime < statB.mtime ? 1 : -1;
			} else {
				return comparers.compareFileNames(statA.name, statB.name);
			}

		case 'filesFirst':
			if (statA.isDirectory && !statB.isDirectory) {
				return 1;
			}
			if (statB.isDirectory && !statA.isDirectory) {
				return -1;
			}
			if (statA instanceof NewStatPlaceholder) {
				return -1;
			}
			if (statB instanceof NewStatPlaceholder) {
				return 1;
			}
			return comparers.compareFileNames(statA.name, statB.name);

		case 'mixed':
			if (statA instanceof NewStatPlaceholder) {
				return -1;
			}
			if (statB instanceof NewStatPlaceholder) {
				return 1;
			}
			return comparers.compareFileNames(statA.name, statB.name);
	}
}
```
Yeah, it looks more verbose but definitely less confusing. Should I commit this?